### PR TITLE
ARROW-2981: [C++] improve clang-tidy usability

### PR DESCRIPTION
**Pull request by bkietz**
- adds a docker-compose service for running clang-tidy
- docker-compose runs as root, so the files touched by clang-tidy and clang-format were owned by root. They are now passed back to the user
- clang-format is run after clang-tidy because the latter munges formatting

I ran clang-tidy then cleaned up the build errors [in this branch](https://github.com/apache/arrow/compare/f92830fa751d791854e9bac9c34755dd730ec375...bkietz:clang-tidy-run-example) to give an idea of what things are changed and what things can go wrong.

### CODE REVIEW

File: .clang-tidy [hunk 1]
```diff
@@ -16,7 +16,7 @@
 # under the License.
 ---
 Checks:          'clang-diagnostic-*,clang-analyzer-*,-clang-analyzer-alpha*,google-*,modernize-*,readability-*'
-HeaderFilterRegex: 'arrow/.*'
+HeaderFilterRegex: '^(.*codegen.*|.*_generated.*|.*windows_compatibility.h|.*pyarrow_api.h|.*pyarrow_lib.h|.*python/config.h|.*python/platform.h|.*thirdparty/ae/.*|.*vendored/.*|.*RcppExports.cpp.*|)$'
```
Comment (fsaintjacques) [inline .clang-tidy]: Add a comment how to reproduce.

File: dev/lint/run_clang_format.sh [hunk 2]
```diff
@@ -18,8 +18,16 @@
...
+  if [ $OWNER ]
+  then
+    chown -R $OWNER /arrow/cpp
+  fi
+}
+
 mkdir -p /build/lint
 pushd /build/lint
   cmake -GNinja /arrow/cpp
+  trap reset_owner EXIT
```
Comment (fsaintjacques) [inline dev/lint/run_clang_format.sh]: This seems fishy. I'd say this is a problem with the docker setup. You can also just run `check-format` for the same purpose without the side effect.

**Review by fsaintjacques — CHANGES_REQUESTED** [review]

File: .clang-tidy (hunk 1 above)
Comment (bkietz) [reply .clang-tidy]: will do

File: dev/lint/run_clang_format.sh (hunk 2 above)
Comment (bkietz) [reply dev/lint/run_clang_format.sh]: I've tried setting [`user: ${UID}`](https://docs.docker.com/compose/compose-file/#domainname-hostname-ipc-mac_address-privileged-read_only-shm_size-stdin_open-tty-user-working_dir) in docker-compose.yml, but the artifacts produced by `docker-compose build cpp` etc are owned by root. I don't know how to pass my uid to `docker-compose build`.

I'm not sure what you mean with `clang-format`; that's an assertion whereas `docker-compose.yml` on master runs [`ninja format`](https://github.com/apache/arrow/blob/master/dev/lint/run_clang_format.sh#L24), which modifies files in place.

Comment (bkietz) [reply dev/lint/run_clang_format.sh]: @fsaintjacques How's this? I create a [user:group](https://github.com/apache/arrow/pull/4293/files#diff-28781500a62df443db3546cfdca67732R24) with the same ids as the host user, then `su` run the linters [with that](https://github.com/apache/arrow/pull/4293/files#diff-2bb746b0d6cff075e06c0bef65542e4cR26)

**Comment by wesm** [comment]
Is the only option to run this and have it apply changes, or is there a way to have it run a report for the purpose of applying manual fixes (similar to IWYU)?

**Comment by bkietz** [comment]
It's not the only option; I could add `format-check` and `clang-tidy-check` endpoints which don't modify. This PR uses the modifying target because that's what `docker-compose run format` does

**Comment by fsaintjacques** [comment]
I think `docker-compose run format` is broken and should not modify the repository volume. It should only return failures and propagate the return code(s). I'd say that running su/sudo in the script blocking.

**Comment by bkietz** [comment]
Okay, I'll refactor both services

**Comment by wesm** [comment]
@fsaintjacques one of the objectives here is to enable auto-formatting to happen in Docker without having to configure clang-format in your local environment. This has been a problem for me when doing Windows development, for example

**Comment by wesm** [comment]
@bkietz can you hold off until we reach some consensus about the behavior of "format"?

**Comment by wesm** [comment]
So the objective of this JIRA is to allow more people to run clang-tidy and look at the output

How about this

* Leave `run format` behavior unchanged -- that's out of scope
* Run clang-tidy and output warnings, but do not modify the codebase

If we want to add a clang-tidy-fix target or something later we can do that.

**Comment by bkietz** [comment]
@wesm done and green

**Comment by fsaintjacques** [comment]
I tried running the checks but it takes way too long. The following changes makes it acceptable, but less robust:
```
-Checks:          'clang-diagnostic-*,clang-analyzer-*,-clang-analyzer-alpha*,google-*,modernize-*,readability-*'
+Checks:          'clang-diagnostic-*,-clang-analyzer-*,-clang-analyzer-alpha*,google-*,modernize-*,readability-*'
```

After this, it seems somthing is wrong:

```
clang-tidy suggested fixes for /home/fsaintjacques/src/db/arrow/cpp/src/arrow/record_batch.cc:
clang-tidy suggested fixes for /home/fsaintjacques/src/db/arrow/cpp/src/arrow/compare.cc:
clang-tidy suggested fixes for /home/fsaintjacques/src/db/arrow/cpp/src/arrow/table_builder.cc:
```

**Review by fsaintjacques — APPROVED** [review]

**Pull request closed without merging**